### DISCUSSION TITLE
Can you compare async with multiple threads?

### DISCUSSION THREAD

**Discussion by pwtail** [psycopg/psycopg#445] [Q&A]
How would you compare the performance of psycopg3 with launching ~2N threads (where N is number of cores, no Python interpreter involved here), each thread holding a connection?

**Comment by dvarrazzo** [comment]
libpq I/O operations should release the GIL, so it should be faster than being serializing the accesses.

**Reply by pwtail** [reply]
You mean, threads should be faster? But no real data? What if there can be multiple async threads as well? Because, for example, there are rust drivers, postgres and tokio-postgres (the latter being async), and tokio-postgres is slower. Both use multiple threads

https://github.com/bikeshedder/rust-postgres-benchmark

**Reply by dvarrazzo** [reply]
I assume that, for what psycopg does, which in a complex program, or in a distributed system, is a little part, threads waiting on I/O shouldn't serialize. This is the only thing that I'm saying.

Then I don't know what the program would do with the rest of its time: is it serving http requests? https? is it multiplying matrices? And what is the database doing? Can the queries that the program is running actually run in parallel or do they get serialized? How many core has the database servers? How many disks? How many other connections?

The fact that, in a different language, a different driver is faster or slower, is nothing you can base a comparison on, or draw conclusions about a program will behave, or how a distributed system of which your program is a small part will behave. I don't have answers out of the box of the type "X is 25% faster than Y".

**Reply by pwtail** [reply]
You may assume that the database size is up to 1GB and the queries are mostly selects (serving web requests, yes)

If neither way is faster by more than 50% on the average the performance can be considered roughly same I think

**Reply by pwtail** [reply]
I've ran a benchmark - asyncpg behaves slightly better than threads given it is using the same number of connections. But I think it is just Cython showing itself, the async nature has little to do with it. The performance boost from using multiple connections is roughly the same for asyncpg and threads.
